dimod/include/dimod/iterators.h

```diff
@@ -0,0 +1,124 @@
+// Copyright 2022 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#pragma once
+
+namespace dimod {
+namespace iterators {
+
+template <class Iter, class T = typename std::iterator_traits<Iter>::value_type::element_type>
+class ConstraintIterator {
+    static_assert(std::is_same<std::random_access_iterator_tag,
+                               typename std::iterator_traits<Iter>::iterator_category>::value,
+                  "template value Iter must be a random access iterator");
+    // Also must be a vector of shared_ptr or unique_ptr, but that's hard to static_assert
+    // and will raise a compiler error regardless, albiet a hard to read one.
+
+    // Without a whole lot of work this could be generalized to being a full indirect iterator,
+    // but it keeps the boilerplate down to be able to make assumptions about the form.
+
+ public:
+    using value_type = T;
+    using pointer = value_type*;
+    using reference = value_type&;
+    using difference_type = typename std::iterator_traits<Iter>::difference_type;
+    using iterator_category = std::random_access_iterator_tag;
+
+    explicit ConstraintIterator(Iter& iterator) : iterator_(iterator) {}
+
+    reference operator*() { return *(*iterator_); }
```

Comment:
Is it trying to create a temporary object and then returning its reference, will it exist after the function quits ?

Reply:
No temporary objects here I don't think? It's the reference to the object pointed to by the iterator.